File: tzone-accounting/frontend/src/core/registry.ts
/**
 * The client kernel: resolve the module graph, run each module's `setup`, and hold what they
 * contributed.
 *
 * Same three passes as the server (`backend/app/core/registry.py`): manifests are declared up
 * front, dependencies are resolved into a deterministic order, then each module registers its
 * tables, screens, menus, translations and hooks. Nothing in this file mentions accounting.
 */

import { HookBus } from "./hooks";
import type {
  DashboardCard,
  EntityDef,
  LocaleBundle,
  MenuItem,
  ModuleContext,
  ModuleManifest,
  RouteDef,
  Seed,
  SettingsPanel,
} from "./types";

export class ModuleError extends Error {}

export class Registry {
  readonly hooks = new HookBus();
  readonly modules = new Map<string, ModuleManifest>();
  readonly entities = new Map<string, EntityDef>();
  readonly localStores = new Map<string, string>();
  readonly routes: RouteDef[] = [];
  readonly menu: MenuItem[] = [];
  readonly seeds: Array<{ module: string; seed: Seed }> = [];
  readonly dashboardCards: DashboardCard[] = [];
  readonly settingsPanels: SettingsPanel[] = [];
  readonly settingsDefaults: Record<string, unknown> = {};
  readonly translations: LocaleBundle = { en: {}, ar: {} };
  installOrder: string[] = [];

  private loading = "core";
  private loaded = false;

  private context(): ModuleContext {
    const registry = this;
    return {
      get key() {
        return registry.loading;
      },
      addEntity(entity) {
        const existing = registry.entities.get(entity.name);
        if (existing) {
          throw new ModuleError(
            `module '${registry.loading}': entity '${entity.name}' is already declared by '${existing.module}'`,
          );
        }
        registry.entities.set(entity.name, { ...entity, module: registry.loading });
      },
      addLocalStore(store, indexes) {
        registry.localStores.set(store, indexes);
      },
      addRoute(route) {
        registry.routes.push(route);
      },
      addMenu(item) {
        registry.menu.push({ sequence: 100, ...item });
      },
      addTranslations(bundle) {
        Object.assign(registry.translations.en, bundle.en);
        Object.assign(registry.translations.ar, bundle.ar);
      },
      addSeed(seed) {
        registry.seeds.push({ module: registry.loading, seed });
      },
      addSettingsDefaults(values) {
        Object.assign(registry.settingsDefaults, values);
      },
      addDashboardCard(card) {
        registry.dashboardCards.push({ sequence: 100, ...card });
      },
      addSettingsPanel(panel) {
        registry.settingsPanels.push({ sequence: 100, ...panel });
      },
      on(hook, handler, sequence) {
        registry.hooks.on(hook, handler, sequence);
      },
    };
  }

  load(manifests: ModuleManifest[], only?: string[]): Registry {
    if (this.loaded) return this;
    const selected = select(manifests, only);
    const context = this.context();

    for (const manifest of resolveOrder(selected)) {
      this.loading = manifest.key;
      this.hooks.bindingModule(manifest.key);
      this.modules.set(manifest.key, manifest);
      manifest.setup(context);
      this.installOrder.push(manifest.key);
    }

    this.loading = "core";
    this.hooks.bindingModule("core");
    this.menu.sort((a, b) => (a.sequence ?? 100) - (b.sequence ?? 100));
    this.dashboardCards.sort((a, b) => (a.sequence ?? 100) - (b.sequence ?? 100));
    this.settingsPanels.sort((a, b) => (a.sequence ?? 100) - (b.sequence ?? 100));
    this.loaded = true;
    return this;
  }

  /** The Dexie schema, assembled from every declared table. */
  schema(): Record<string, string> {
    const stores: Record<string, string> = {};
    for (const entity of this.entities.values()) stores[entity.store] = entity.indexes;
    for (const [store, indexes] of this.localStores) stores[store] = indexes;
    return stores;
  }

  entityForStore(store: string): EntityDef | undefined {
    for (const entity of this.entities.values()) {
      if (entity.store === store) return entity;
    }
    return undefined;
  }

  describe() {
    return {
      installOrder: this.installOrder,
      modules: [...this.modules.values()].map((m) => ({
        key: m.key,
        name: m.name,
        nameAr: m.nameAr,
        version: m.version,
        summary: m.summary,
        category: m.category,
        depends: m.depends,
        entities: [...this.entities.values()]
          .filter((e) => e.module === m.key)
          .map((e) => e.name),
        screens: this.routes.length,
      })),
      hooks: this.hooks.describe(),
    };
  }
}

function select(manifests: ModuleManifest[], only?: string[]): ModuleManifest[] {
  const byKey = new Map(manifests.map((m) => [m.key, m]));
  if (!only) return manifests;

  const wanted = new Map<string, ModuleManifest>();
  const stack = [...only];
  while (stack.length) {
    const key = stack.pop()!;
    if (wanted.has(key)) continue;
    const manifest = byKey.get(key);
    if (!manifest) throw new ModuleError(`unknown module '${key}'`);
    wanted.set(key, manifest);
    stack.push(...manifest.depends);
  }
  return [...wanted.values()];
}

/** Topological sort; ties broken by `sequence` then key so the order is machine-independent. */
export function resolveOrder(manifests: ModuleManifest[]): ModuleManifest[] {
  const byKey = new Map(manifests.map((m) => [m.key, m]));
  for (const manifest of manifests) {
    for (const dependency of manifest.depends) {
      if (!byKey.has(dependency)) {
        throw new ModuleError(
          `module '${manifest.key}' depends on '${dependency}', which is not installed`,
        );
      }
    }
  }

  const ordered: ModuleManifest[] = [];
  const placed = new Set<string>();
  const remaining = new Map(byKey);

  while (remaining.size) {
    const ready = [...remaining.values()].filter((m) =>
      m.depends.every((d) => placed.has(d)),
    );
    if (!ready.length) {
      throw new ModuleError(
        `dependency cycle between modules: ${[...remaining.keys()].sort().join(", ")}`,
      );
    }
    ready.sort((a, b) => (a.sequence ?? 100) - (b.sequence ?? 100) || a.key.localeCompare(b.key));
    for (const manifest of ready) {
      ordered.push(manifest);
      placed.add(manifest.key);
      remaining.delete(manifest.key);
    }
  }
  return ordered;
}

let instance: Registry | null = null;

export function getRegistry(): Registry {
  if (!instance) throw new ModuleError("registry not installed yet — call installModules() first");
  return instance;
}

export function installModules(manifests: ModuleManifest[], only?: string[]): Registry {
  instance = new Registry().load(manifests, only);
  return instance;
}

export function resetRegistry(): void {
  instance = null;
}

File: tzone-accounting/backend/app/core/registry.py
"""The kernel: discover modules, order them, load them, and hold what they contribute.

Loading is three passes, in Odoo's spirit:

1. **Discover** — read every `app/modules/*/__manifest__.py`. Manifests are data, so the
   dependency graph is known before any module code runs.
2. **Resolve** — topological sort over `depends`, ties broken by `sequence` then key, so the
   install order is deterministic on every machine.
3. **Load** — import each module's `module.py` and call its `setup(registry)`, which is the one
   and only place a module registers tables, entities, routers, seeds and hooks.

Nothing in this file mentions accounting. Deleting every module directory would leave a working
(if empty) server.
"""

from __future__ import annotations

import importlib
import os
import sqlite3
from pathlib import Path
from typing import Any, Callable

from fastapi import APIRouter

from .entities import EntityDescriptor
from .errors import ModuleError
from .hooks import HookBus
from .manifest import LoadedModule, Manifest

MODULES_DIR = Path(__file__).resolve().parent.parent / "modules"
MODULES_PACKAGE = "app.modules"

SeedFn = Callable[[sqlite3.Connection], None]


class Registry:
    """Everything the running server knows, contributed by modules."""

    def __init__(self) -> None:
        self.hooks = HookBus()
        self.modules: dict[str, LoadedModule] = {}
        self.entities: dict[str, EntityDescriptor] = {}
        self.routers: list[APIRouter] = []
        self.schema_fragments: list[tuple[str, str]] = []
        self.seeds: list[tuple[str, SeedFn]] = []
        self.settings_defaults: dict[str, Any] = {}
        self._loading: str = "core"
        self._loaded = False

    # ---------------------------------------------------------------- module-facing API

    def add_schema(self, sql: str) -> None:
        """Contribute a schema fragment. Must be idempotent (`CREATE TABLE IF NOT EXISTS`)."""
        self.schema_fragments.append((self._loading, sql))

    def add_entity(self, descriptor: EntityDescriptor) -> None:
        """Declare a replicated table. It joins the sync protocol with no further work."""
        if descriptor.name in self.entities:
            raise ModuleError(
                f"module {self._loading!r}: entity {descriptor.name!r} is already declared by "
                f"{self.entities[descriptor.name].module!r}"
            )
        descriptor.module = self._loading
        self.entities[descriptor.name] = descriptor

    def extend_entity(
        self,
        name: str,
        *,
        validators: list = None,
        before_write: list = None,
    ) -> None:
        """Attach behaviour to an entity another module owns.

        This is how `invoicing` puts legal numbering on `documents`' table without either
        module importing the other.
        """
        descriptor = self.entities.get(name)
        if descriptor is None:
            raise ModuleError(
                f"module {self._loading!r}: cannot extend unknown entity {name!r}"
            )
        descriptor.validators.extend(validators or [])
        descriptor.before_write.extend(before_write or [])

    def add_router(self, router: APIRouter) -> None:
        self.routers.append(router)
        self.modules[self._loading].routers += 1

    def add_seed(self, fn: SeedFn) -> None:
        """Data every fresh database needs. Seeds run in module install order, idempotently."""
        self.seeds.append((self._loading, fn))

    def add_settings_defaults(self, values: dict[str, Any]) -> None:
        """Contribute keys to the company settings document."""
        self.settings_defaults.update(values)

    def on(self, hook: str, fn: Callable[..., Any], sequence: int = 10) -> None:
        self.hooks.on(hook, fn, sequence)

    # ---------------------------------------------------------------- loading

    def load(self, only: list[str] | None = None) -> "Registry":
        if self._loaded:
            return self
        manifests = discover(MODULES_DIR)
        selected = _select(manifests, only)
        for manifest in resolve_order(selected):
            self._loading = manifest.key
            self.hooks.binding_module(manifest.key)
            self.modules[manifest.key] = LoadedModule(manifest=manifest)
            module = importlib.import_module(f"{MODULES_PACKAGE}.{manifest.key}.module")
            setup = getattr(module, "setup", None)
            if setup is None:
                raise ModuleError(f"module {manifest.key!r}: module.py has no setup(registry)")
            setup(self)
            self.modules[manifest.key].entities = [
                name for name, e in self.entities.items() if e.module == manifest.key
            ]
        self._loading = "core"
        self.hooks.binding_module("core")
        self._loaded = True
        return self

    def schema(self) -> str:
        return "\n\n".join(
            f"-- module: {key}\n{sql}" for key, sql in self.schema_fragments
        )

    def describe(self) -> dict:
        return {
            "modules": [
                {
                    **loaded.manifest.as_dict(),
                    "entities": loaded.entities,
                    "routers": loaded.routers,
                }
                for loaded in self.modules.values()
            ],
            "entities": sorted(self.entities),
            "hooks": self.hooks.describe(),
        }


# ---------------------------------------------------------------------- discovery


def discover(root: Path) -> dict[str, Manifest]:
    manifests: dict[str, Manifest] = {}
    if not root.exists():
        return manifests
    for entry in sorted(root.iterdir()):
        if not entry.is_dir() or entry.name.startswith((".", "_")):
            continue
        if not (entry / "__manifest__.py").exists():
            continue
        module = importlib.import_module(f"{MODULES_PACKAGE}.{entry.name}.__manifest__")
        data = getattr(module, "MANIFEST", None)
        if data is None:
            raise ModuleError(f"module {entry.name!r}: __manifest__.py defines no MANIFEST")
        manifests[entry.name] = Manifest.from_dict(entry.name, data)
    return manifests


def _select(manifests: dict[str, Manifest], only: list[str] | None) -> dict[str, Manifest]:
    """Pick the modules to install, pulling in dependencies of anything explicitly selected."""
    available = {k: m for k, m in manifests.items() if m.installable}
    if only is None:
        env = os.environ.get("ACCOUNTING_MODULES", "").strip()
        only = [part.strip() for part in env.split(",") if part.strip()] or None
    if only is None:
        return available

    wanted: dict[str, Manifest] = {}
    stack = list(only)
    while stack:
        key = stack.pop()
        if key in wanted:
            continue
        manifest = available.get(key)
        if manifest is None:
            raise ModuleError(f"unknown or non-installable module {key!r}")
        wanted[key] = manifest
        stack.extend(manifest.depends)

    # auto_install modules join once all of their dependencies are in.
    changed = True
    while changed:
        changed = False
        for key, manifest in available.items():
            if key in wanted or not manifest.auto_install:
                continue
            if all(dep in wanted for dep in manifest.depends):
                wanted[key] = manifest
                changed = True
    return wanted


def resolve_order(manifests: dict[str, Manifest]) -> list[Manifest]:
    """Topological sort. Raises on a missing dependency or a cycle, naming the modules."""
    for manifest in manifests.values():
        for dependency in manifest.depends:
            if dependency not in manifests:
                raise ModuleError(
                    f"module {manifest.key!r} depends on {dependency!r}, which is not installed"
                )

    ordered: list[Manifest] = []
    placed: set[str] = set()
    remaining = dict(manifests)

    while remaining:
        ready = [
            manifest
            for manifest in remaining.values()
            if all(dependency in placed for dependency in manifest.depends)
        ]
        if not ready:
            raise ModuleError(
                "dependency cycle between modules: " + ", ".join(sorted(remaining))
            )
        ready.sort(key=lambda m: (m.sequence, m.key))
        for manifest in ready:
            ordered.append(manifest)
            placed.add(manifest.key)
            del remaining[manifest.key]
    return ordered


_registry: Registry | None = None


def get_registry() -> Registry:
    """The process-wide registry, loaded on first use."""
    global _registry
    if _registry is None:
        _registry = Registry().load()
    return _registry


def reset_registry() -> None:
    """Drop the loaded registry. Used by tests that install a different module set."""
    global _registry
    _registry = None
